[containerapp] Enable create/up w --repo without Dockerfile

snehapar9 opened the pull request:
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
<!--- Please provide the related command with az {command} if you can, so that we can quickly route to the related person to review. --->


### General Guidelines

- [x] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).


### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

yonzhan commented:
<!-- add-pr-comment:add-pr-comment -->

Thank you for your contribution! We will review the pull request and get back to you soon.

cormacpayne commented on src/containerapp/azext_containerapp/containerapp_decorator.py:
```diff
@@ -1294,7 +1294,7 @@ def _construct_app_and_env_for_source_or_repo(self):
         image = None if self.get_argument_image().__eq__(HELLO_WORLD_IMAGE) else _reformat_image(self.get_argument_source(), self.get_argument_repo(), self.get_argument_image())
 
         has_dockerfile = _has_dockerfile(self.get_argument_source(), dockerfile)
-        if not self.get_argument_source() or has_dockerfile:
```
Would you mind explaining how the changes made in this PR are enabling these commands to build a provided GitHub repository without a Dockerfile?

daniv-msft replied:
+1, I would appreciate a bit more explanation as well as the changes are not super self-evident.

snehapar9 replied:
Thank you @daniv-msft  and @cormacpayne for taking a look! 
As of today, az containerapp up/create w --repo throws a validation error if there was no Dockerfile in the repo [here](https://github.com/snehapar9/azure-cli-extensions/blob/main/src/containerapp/azext_containerapp/_up_utils.py#L718)

Removed the condition to check for `not self.get_argument_source()` i.e. `repo` from this line because it is not necessary to check if the customer's repo contains a Dockerfile here. The `container-app-deploy-action` looks for `Dockerfile` by default first before attempting to use the builder to build the app.

But, if folks feel we should check if repo contains `Dockerfile` here, without solely relying on the GitHub action, I can modify `_has_dockerfile` to  determine if either source or repo contains `Dockerfile` and remove [this validation](https://github.com/snehapar9/azure-cli-extensions/blob/main/src/containerapp/azext_containerapp/_up_utils.py#L718) validation instead.

cormacpayne replied:
OK I'm just now looking up the [`_get_dockerfile_content`](https://github.com/Azure/azure-cli-extensions/blob/main/src/containerapp/azext_containerapp/_up_utils.py#L802) function and this change makes sense now.

Previously, if no local application source was provided, we would still meet this conditional and call the `_get_dockerfile_content` function, and once inside that function, we would've hit the repository conditional since we could have _only_ called this function if the user provided the corresponding source or repository argument.

Because our GitHub Action is going to check for the Dockerfile already, we don't need additional logic in the code to make decisions that are dependent on the repository having a Dockerfile already. I think it makes sense to explicitly handle the local source with Dockerfile case using the result of `_has_dockerfile` and deferring any repository Dockerfile case to the GitHub Action itself.

I would personally leave a comment above the modified lines to signal that we only care about the local source Dockerfile cases and can safely ignore the Dockerfile cases in the repository cases due to its inherent check of the Dockerfile (however you'd like to word that).

snehapar9 replied:
Added comment to justify why we don't need to look for Dockerfile explicitly in case of `--repo`

cormacpayne reviewed (approved):
:shipit:

Greedygre commented on src/containerapp/azext_containerapp/containerapp_decorator.py:
```diff
@@ -1294,7 +1294,7 @@ def _construct_app_and_env_for_source_or_repo(self):
         image = None if self.get_argument_image().__eq__(HELLO_WORLD_IMAGE) else _reformat_image(self.get_argument_source(), self.get_argument_repo(), self.get_argument_image())
```
Hi @snehapar9 
Please add a description for this change in history.rst.

Greedygre replied:
Could you please rerun tests in `test_containerapp_create_update_with_source.py` and update the recording file?

snehapar9 replied:
Updated.

Greedygre commented:
Hi @snehapar9 
Please rerun tests in `test_containerapp_create_update_with_source.py` and update the recording files.
Then you can update the version to `0.3.41` in `set_up.py` and `history.rst` to release a new version, sample pr: https://github.com/Azure/azure-cli-extensions/pull/6619/files

Generally, we need to re-run all tests before release a new extension version. As I just rerun all test 2 days ago, and there is only this PR change, so no need to rerun all tests again.

snehapar9 commented:
> Hi @snehapar9 Please rerun tests in `test_containerapp_create_update_with_source.py` and update the recording files. Then you can update the version to `0.3.41` in `set_up.py` and `history.rst` to release a new version, sample pr: https://github.com/Azure/azure-cli-extensions/pull/6619/files
> 
> Generally, we need to re-run all tests before release a new extension version. As I just rerun all test 2 days ago, and there is only this PR change, so no need to rerun all tests again.

Addressed PR comments. Here are successful workflow runs for with and without Dockerfile - 
1. Dotnet App without Dockerfile - https://github.com/snehapar9/HelloWorldApp/actions/runs/6271005127
2. Node js App with Dockerfile - https://github.com/snehapar9/nextjs-basic/actions/runs/6270199216

Greedygre commented on src/containerapp/HISTORY.rst:
```diff
@@ -3,7 +3,9 @@
 Release History
 ===============
 upcoming
```
```suggestion
upcoming

```

Greedygre commented on src/containerapp/HISTORY.rst:
```diff
@@ -3,7 +3,9 @@
 Release History
 ===============
 upcoming
+0.3.41
```
please change version in `set_up.py`.

daniv-msft commented on src/containerapp/azext_containerapp/custom.py:
```diff
@@ -1958,7 +1958,7 @@ def create_or_update_github_action(cmd,
 
             raise ValidationError(
-                "Failed to find workflow file for Container App '{}' in .github/workflow folder for repo '{}'. ".format(name, repo) +
+                "Failed to find workflow file for Container App '{}' in .github/workflow folder for repo '{}'. ".format(name, repo) + "Please try again by re-running the command."
```
Would we need to end this line by a "+" so that the string can be concatenated with the next error string?
Also, should we directly put the new error string with the previous one rather than concatenating it?

Something like:
```
"Failed to find workflow file for Container App '{}' in .github/workflow folder for repo '{}'. Please try again by re-running the command. ".format(name, repo) +
```

cormacpayne reviewed (approved):
:shipit:

snehapar9 commented on src/containerapp/azext_containerapp/custom.py:
```diff
@@ -1958,8 +1958,8 @@ def create_or_update_github_action(cmd,
             raise ValidationError(
-                "Failed to find workflow file for Container App '{}' in .github/workflow folder for repo '{}'. ".format(name, repo) +
-                "If this file was removed, please use the 'az containerapp github-action delete' command to disconnect the removed workflow file connection.")
+                "Exhausted the number of re-tries allotted to polling the creation of the workflow file for Container App '{}' in .github/workflow folder for repo '{}'. ".format(name, repo) +
```
This change was made to make sure we accurately let the customer know that the operation to get the workflow file created did not success after 6 re-tries but it's likely that the customer would see it in their repository after the create operation finishes.

Greedygre commented:
Hi @yanzhudd 
Could you please help to review this PR? Thanks.

yanzhudd commented on src/containerapp/HISTORY.rst:
```diff
@@ -3,7 +3,10 @@
+
+0.3.41
 ++++++
+* 'az containerapp create': enable support for no Dockerfile cases with --repo
```
May I ask if this change only applies to `create` command not including `update` command?

Greedygre replied:
```suggestion
* 'az containerapp up/create': enable support for no Dockerfile cases with --repo
```

yanzhudd commented:
Please rename the title.

Greedygre commented:
> Please rename the title.

Recommend:
[containerapp] `az containerapp up/create`: [Enable create/up w --repo without Dockerfile](https://github.com/Azure/azure-cli-extensions/pull/6785#top)

Greedygre commented:
Hi @snehapar9 

For command `az containerapp up`, as we have the same command and feature in Azure-cli, please fix in https://github.com/Azure/azure-cli/tree/dev/src/azure-cli/azure/cli/command_modules/containerapp too.
